File: src/main/java/Q59_SpiralMatrixII.java
/***
 * 44mins medium small bug
 * Runtime: 0 ms, faster than 100.00% of Java online submissions for Spiral Matrix II.
 * Memory Usage: 37.1 MB, less than 87.40% of Java online submissions for Spiral Matrix II.
 *
 */
public class Q59_SpiralMatrixII {
    public int[][] generateMatrix(int n) {
        int[][] result = new int[n][n];
        int value = 1;
        int layer = 0;
        int i = 0;
        int j = 0;
        int left = layer;
        int right = n - layer;
        while(value <= (n * n)) {
            result[i][j] = value++;
            if(i == left) {
                if(j < right - 1) {
                    j++;
                }
                else {
                    i++;
                }
            }
            else if(j == right - 1) {
                if(i < right - 1) {
                    i++;
                }
                else {
                    j--;
                }
            }
            else if(i == right - 1) {
                if(j > left) {
                    j--;
                }
                else {
                    i--;
                }
            }
            else if(j == left) {
                if(i > left) {
                    i--;
                }
            }

            if(i == left && j == left) {
                layer++;
                left = layer;
                right = n - layer;
                i = left;
                j = left;
            }

        }
        return result;
    }
}

File: LeetCodePractice/LeetCodePractice/Q59_SpiralMatrixII.cpp
#include "LCHeader.h"
//3:09 3:23
/*
Runtime: 0 ms, faster than 100.00% of C++ online submissions for Spiral Matrix II.
Memory Usage: 6.6 MB, less than 51.28% of C++ online submissions for Spiral Matrix II.
*/
class Solution {//3:09 3:23
public:
    vector<vector<int>> generateMatrix(int n) {
        vector<vector<int>> result(n, vector<int>(n, 0));
        int count = n * n;
        int idx = 0;
        vector<vector<int>*> step;
        vector<int> start{ 0, -1 };
        int N = n + 2;
        while (idx < count) {
            if (step.size() == 0) {
                start[1]++;
                result[start[0]][start[1]] = idx + 1;
                idx++;
                N -= 2;
                step = comp(N);
                continue;
            }
            for (int i = 0; i < step.size(); i++) {
                start[0] += (*step[i])[0];
                start[1] += (*step[i])[1];
                result[start[0]][start[1]] = idx + 1;
                idx++;
            }
            step.clear();
        }
        return result;
    }

    vector<vector<int>*> comp(int n) {
        vector<vector<int>*> result;
        static vector<int> v1{ 0, 1 };
        static vector<int> v2{ 1, 0 };
        static vector<int> v3{ 0, -1 };
        static vector<int> v4{ -1, 0 };
        for (int i = 0; i < n - 1; i++) {
            result.push_back(&v1);
        }
        for (int i = 0; i < n - 1; i++) {
            result.push_back(&v2);
        }
        for (int i = 0; i < n - 1; i++) {
            result.push_back(&v3);
        }
        for (int i = 0; i < n - 2; i++) {
            result.push_back(&v4);
        }
        return result;
    }

};